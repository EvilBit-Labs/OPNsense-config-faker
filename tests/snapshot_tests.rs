//! Snapshot tests for OPNsense Config Faker
//!
//! These tests capture output snapshots to ensure CLI output remains consistent
//! across changes and updates.

mod common;

use common::{cli_command, create_temp_csv, create_temp_xml, normalize_output, TestOutputExt};
use insta::assert_snapshot;

/// Test CSV generation output with stable snapshots
#[test]
fn test_csv_output_snapshot() {
    let (temp_file, csv_path) = create_temp_csv(
        "snapshot_test_",
        &[
            &["VLAN", "IP Range", "Description"],
            &["100", "192.168.1.0/24", "Test Network"],
            &["200", "192.168.2.0/24", "Production"],
        ],
    )
    .unwrap();

    let output = cli_command()
        .arg("generate")
        .arg("--format")
        .arg("csv")
        .arg("--count")
        .arg("3")
        .arg("--output")
        .arg(&csv_path)
        .arg("--force") // Force overwrite existing temp file
        .arg("--seed")
        .arg("12345") // Fixed seed for reproducible output
        .run_success();

    // Normalize output for stable snapshots
    let normalized_stdout = normalize_output(&output.stdout);
    let normalized_stderr = normalize_output(&output.stderr);

    // Snapshot the normalized CLI output
    assert_snapshot!("csv_generation_success_stdout", normalized_stdout);
    assert_snapshot!("csv_generation_success_stderr", normalized_stderr);

    // Also snapshot the generated CSV content
    let csv_content = std::fs::read_to_string(&csv_path).unwrap();
    assert_snapshot!("generated_csv_content", csv_content);

    drop(temp_file);
}

/// Test help output snapshots
#[test]
fn test_help_output_snapshots() {
    // Test main help
    let output = cli_command().arg("--help").run_success();

    assert_snapshot!("main_help_output", output.normalized_stdout());

    // Test generate help
    let output = cli_command().arg("generate").arg("--help").run_success();

    assert_snapshot!("generate_help_output", output.normalized_stdout());

    // Test completions help
    let output = cli_command().arg("completions").arg("--help").run_success();

    assert_snapshot!("completions_help_output", output.normalized_stdout());
}

/// Test error output snapshots for consistent error messages
#[test]
fn test_error_output_snapshots() {
    // Test missing output file error
    let output = cli_command()
        .arg("generate")
        .arg("--format")
        .arg("csv")
        .arg("--count")
        .arg("5")
        .run_failure();

    assert_snapshot!("missing_csv_output_error", output.normalized_stderr());

    // Test missing XML base config error
    let output = cli_command()
        .arg("generate")
        .arg("--format")
        .arg("xml")
        .arg("--count")
        .arg("5")
        .run_failure();

    assert_snapshot!("missing_xml_base_config_error", output.normalized_stderr());

    // Test file exists without force error
    let (temp_file, csv_path) = create_temp_csv("error_test_", &[&["existing", "data"]]).unwrap();

    let output = cli_command()
        .arg("generate")
        .arg("--format")
        .arg("csv")
        .arg("--count")
        .arg("2")
        .arg("--output")
        .arg(&csv_path)
        .arg("--seed")
        .arg("42")
        .run_failure();

    assert_snapshot!(
        "file_exists_without_force_error",
        output.normalized_stderr()
    );

    drop(temp_file);
}

/// Test deprecated command warnings
#[test]
fn test_deprecated_command_snapshots() {
    let (temp_file, csv_path) =
        create_temp_csv("deprecated_test_", &[&["header1", "header2"]]).unwrap();

    // Test deprecated CSV command
    let output = cli_command()
        .arg("csv")
        .arg("--count")
        .arg("3")
        .arg("--output")
        .arg(&csv_path)
        .arg("--seed")
        .arg("42")
        .run_failure();

    // Capture both stdout and stderr for deprecation messages
    assert_snapshot!("deprecated_csv_stdout", output.normalized_stdout());
    assert_snapshot!("deprecated_csv_stderr", output.normalized_stderr());

    drop(temp_file);
}

/// Test XML generation output snapshots
#[test]
fn test_xml_generation_snapshots() {
    let base_xml_content = r#"<?xml version="1.0"?>
<opnsense version="24.1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <system>
    <hostname>opnsense</hostname>
  </system>
  <interfaces>
    <lan>
      <if>em0</if>
      <ipaddr>dhcp</ipaddr>
    </lan>
    <wan>
      <if>em1</if>
      <ipaddr>dhcp</ipaddr>
    </wan>
  </interfaces>
</opnsense>"#;

    let (temp_xml_file, xml_path) = create_temp_xml("base_config_", base_xml_content).unwrap();
    let (temp_csv_file, csv_path) = create_temp_csv(
        "vlan_data_",
        &[
            &["VLAN", "IP Range", "Beschreibung", "WAN"],
            &["100", "192.168.100.0/24", "Development", "1"],
            &["200", "192.168.200.0/24", "Testing", "2"],
        ],
    )
    .unwrap();

    // Test XML generation using CSV input
    let output = cli_command()
        .arg("generate")
        .arg("--format")
        .arg("xml")
        .arg("--base-config")
        .arg(&xml_path)
        .arg("--csv-file")
        .arg(&csv_path)
        .arg("--output-dir")
        .arg("/tmp") // Ensure consistent output directory for snapshots
        .arg("--force") // Force overwrite existing files
        .run_success();

    assert_snapshot!("xml_generation_with_csv_stdout", output.normalized_stdout());
    assert_snapshot!("xml_generation_with_csv_stderr", output.normalized_stderr());

    drop(temp_xml_file);
    drop(temp_csv_file);
}

/// Test progress output normalization
#[test]
fn test_progress_normalization() {
    let (temp_file, csv_path) = create_temp_csv("progress_test_", &[]).unwrap();

    // Remove temp file first to avoid force requirement
    drop(temp_file);
    std::fs::remove_file(&csv_path).ok(); // Ignore error if file doesn't exist

    let output = cli_command()
        .arg("generate")
        .arg("--format")
        .arg("csv")
        .arg("--count")
        .arg("10") // Larger count to trigger progress indicators
        .arg("--output")
        .arg(&csv_path)
        .arg("--seed")
        .arg("999")
        .run_success();

    // The normalize_output function should remove progress indicators
    let normalized = normalize_output(&output.stdout);

    // Verify progress characters are removed
    assert!(!normalized.contains("⠁"));
    assert!(!normalized.contains("⠂"));
    assert!(!normalized.contains("⠄"));
    assert!(!normalized.contains("▪"));
    assert!(!normalized.contains("▫"));

    assert_snapshot!("progress_normalized_output", normalized);
}

/// Test bash completion script snapshot
#[test]
fn test_bash_completion_snapshot() {
    let output = cli_command().arg("completions").arg("bash").run_success();

    // Bash completions should be consistent
    let normalized = normalize_output(&output.stdout);
    assert_snapshot!("bash_completion_script", normalized);
}

/// Test zsh completion script snapshot
#[test]
fn test_zsh_completion_snapshot() {
    let output = cli_command().arg("completions").arg("zsh").run_success();

    let normalized = normalize_output(&output.stdout);
    assert_snapshot!("zsh_completion_script", normalized);
}

/// Test fish completion script snapshot
#[test]
fn test_fish_completion_snapshot() {
    let output = cli_command().arg("completions").arg("fish").run_success();

    let normalized = normalize_output(&output.stdout);
    assert_snapshot!("fish_completion_script", normalized);
}
