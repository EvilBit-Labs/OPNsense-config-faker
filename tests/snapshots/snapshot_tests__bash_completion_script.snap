---
source: tests/snapshot_tests.rs
expression: normalized
---
_opnsense-config-faker() { local i cur prev opts cmd COMPREPLY=() if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then cur="$2" else cur="${COMP_WORDS[COMP_CWORD]}" fi prev="$3" cmd="" opts="" for i in "${COMP_WORDS[@]:0:COMP_CWORD}" do case "${cmd},${i}" in ",$1") cmd="opnsense__config__faker" ;; opnsense__config__faker,completions) cmd="opnsense__config__faker__completions" ;; opnsense__config__faker,csv) cmd="opnsense__config__faker__csv" ;; opnsense__config__faker,generate) cmd="opnsense__config__faker__generate" ;; opnsense__config__faker,help) cmd="opnsense__config__faker__help" ;; opnsense__config__faker,xml) cmd="opnsense__config__faker__xml" ;; opnsense__config__faker__help,completions) cmd="opnsense__config__faker__help__completions" ;; opnsense__config__faker__help,csv) cmd="opnsense__config__faker__help__csv" ;; opnsense__config__faker__help,generate) cmd="opnsense__config__faker__help__generate" ;; opnsense__config__faker__help,help) cmd="opnsense__config__faker__help__help" ;; opnsense__config__faker__help,xml) cmd="opnsense__config__faker__help__xml" ;; *) ;; esac done case "${cmd}" in opnsense__config__faker) opts="-h -V --help --version generate completions csv xml help" if [[ ${cur} == -* || ${COMP_CWORD} -eq 1 ]] ; then COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 fi case "${prev}" in *) COMPREPLY=() ;; esac COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 ;; opnsense__config__faker__completions) opts="-h --help bash zsh fish power-shell elvish" if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 fi case "${prev}" in *) COMPREPLY=() ;; esac COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 ;; opnsense__config__faker__csv) opts="-c -o -f -h --count --output --force --seed --help" if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 fi case "${prev}" in --count) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; -c) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --output) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; -o) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --seed) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; *) COMPREPLY=() ;; esac COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 ;; opnsense__config__faker__generate) opts="-f -c -o -b -F -i -h --format --count --output --output-dir --base-config --csv-file --firewall-nr --opt-counter --force --seed --no-color --interactive --help" if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 fi case "${prev}" in --format) COMPREPLY=($(compgen -W "csv xml" -- "${cur}")) return 0 ;; -f) COMPREPLY=($(compgen -W "csv xml" -- "${cur}")) return 0 ;; --count) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; -c) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --output) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; -o) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --output-dir) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --base-config) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; -b) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --csv-file) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --firewall-nr) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --opt-counter) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --seed) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; *) COMPREPLY=() ;; esac COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 ;; opnsense__config__faker__help) opts="generate completions csv xml help" if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 fi case "${prev}" in *) COMPREPLY=() ;; esac COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 ;; opnsense__config__faker__help__completions) opts="" if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 fi case "${prev}" in *) COMPREPLY=() ;; esac COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 ;; opnsense__config__faker__help__csv) opts="" if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 fi case "${prev}" in *) COMPREPLY=() ;; esac COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 ;; opnsense__config__faker__help__generate) opts="" if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 fi case "${prev}" in *) COMPREPLY=() ;; esac COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 ;; opnsense__config__faker__help__help) opts="" if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 fi case "${prev}" in *) COMPREPLY=() ;; esac COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 ;; opnsense__config__faker__help__xml) opts="" if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 fi case "${prev}" in *) COMPREPLY=() ;; esac COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 ;; opnsense__config__faker__xml) opts="-b -c -o -f -h --base-config --count --csv-file --output-dir --firewall-nr --opt-counter --force --seed --help" if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 fi case "${prev}" in --base-config) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; -b) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --count) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; -c) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --csv-file) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --output-dir) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; -o) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --firewall-nr) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --opt-counter) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; --seed) COMPREPLY=($(compgen -f "${cur}")) return 0 ;; *) COMPREPLY=() ;; esac COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") ) return 0 ;; esac } if [[ "${BASH_VERSINFO[0]}" -eq 4 && "${BASH_VERSINFO[1]}" -ge 4 || "${BASH_VERSINFO[0]}" -gt 4 ]]; then complete -F _opnsense-config-faker -o nosort -o bashdefault -o default opnsense-config-faker else complete -F _opnsense-config-faker -o bashdefault -o default opnsense-config-faker fi
